File: java/api/src/main/java/org/dressdiscover/api/models/collection/CollectionId.java
package org.dressdiscover.api.models.collection;

import static com.google.common.base.Preconditions.checkNotNull;

import org.dressdiscover.api.models.institution.InstitutionId;
import org.thryft.waf.api.models.StringModelId;

import org.dressdiscover.api.models.collection.InvalidCollectionIdException;
import org.dressdiscover.api.models.institution.InvalidInstitutionIdException;

public final class CollectionId extends StringModelId {
    public static CollectionId parse(final String value) throws InvalidCollectionIdException {
        if (value.isEmpty()) {
            throw InvalidCollectionIdException.create("empty string");
        }
        final String[] valueSplit = value.split("/");
        if (valueSplit.length != 2) {
            throw InvalidCollectionIdException.create("malformed collection ID: " + value);
        }
        InstitutionId institutionId;
        try {
            institutionId = InstitutionId.parse(valueSplit[0].trim());
        } catch (final InvalidInstitutionIdException e) {
            throw InvalidCollectionIdException.create(e.toString());
        }
        final String unqualifiedCollectionId = valueSplit[1].trim();
        if (unqualifiedCollectionId.isEmpty()) {
            throw InvalidCollectionIdException.create("empty collection ID: " + value);
        }
        return new CollectionId(unqualifiedCollectionId, institutionId, value);
    }

    private CollectionId(final String unqualifiedCollectionId, final InstitutionId institutionId, final String value) {
        super(value);
        this.institutionId = checkNotNull(institutionId);
        this.unqualifiedCollectionId = checkNotNull(unqualifiedCollectionId);
    }

    public InstitutionId getInstitutionId() {
        return institutionId;
    }

    public String getUnqualifiedCollectionId() {
        return unqualifiedCollectionId;
    }

    private final InstitutionId institutionId;
    private final String unqualifiedCollectionId;
}

File: py/src/dressdiscover/api/models/collection/collection_id.py
from dressdiscover.api.models.collection.invalid_collection_id_exception import InvalidCollectionIdException
from dressdiscover.api.models.institution.institution_id import InstitutionId
from dressdiscover.api.models.institution.invalid_institution_id_exception import InvalidInstitutionIdException


class CollectionId(str):
    @classmethod
    def parse(cls, str_):
        if len(str_) == 0:
            raise InvalidCollectionIdException("empty")
        split = str_.split('/')
        if len(split) != 2:
            raise InvalidCollectionIdException("malformed collection ID: " + str_)
        try:
            institution_id = InstitutionId.parse(split[0].strip())
        except InvalidInstitutionIdException, e:
            raise InvalidCollectionIdException(str(e))
        unqualified_collection_id = split[1].strip()
        if len(unqualified_collection_id) == 0:
            raise InvalidCollectionIdException("empty collection id")
        parsed = cls(str_)
        parsed.__institution_id = institution_id
        parsed.__unqualified_collection_id = unqualified_collection_id
        return parsed

    @property
    def institution_id(self):
        return self.__institution_id

    @property
    def unqualified_collection_id(self):
        return self.__unqualified_collection_id

